File: backend/app/schemas/template_column.py
"""
Pydantic schemas for TemplateColumn resources.

Columns are created as a batch for a template - not individually.
So the Create schema represents a single column, and the
endpoint accepts a list of these.
"""

from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import (
    ORMBase,
    IdentifierStr,
    DescriptionStr,
)


# Allowed Unity Catalog data types for template columns.
# Matches the subset of types we support in the DDL builder.
ALLOWED_DATA_TYPES = r"^(STRING|INTEGER|BIGINT|DOUBLE|DECIMAL|BOOLEAN|DATE|TIMESTAMP)$"


class TemplateColumnCreate(BaseModel):
    """
    Request body for creating a single column within a template.

    POST /templates/{id}/columns accepts a list of these.
    """

    column_name: IdentifierStr = Field(
        ...,
        description="Technical column name. Must match file header exactly."
    )
    display_name: Optional[str] = Field(
        None,
        max_length=100,
        description="Human-readable column name for the UI."
    )
    data_type: str = Field(
        ...,
        pattern=ALLOWED_DATA_TYPES,
        description="Unity Catalog data type. Must be one of the supported types."
    )
    description: Optional[DescriptionStr] = None
    is_included: bool = Field(
        True,
        description="Whether this column is included in the target table."
    )
    is_pii: bool = Field(
        False,
        description="Mark as PII to apply column masking."
    )
    is_nullable: bool = Field(
        True,
        description="Whether NULL values are allowed."
    )
    is_unique: bool = Field(
        False,
        description="Enforce uniqueness. Uniqueness is enforced at the Polars validation layer, not Delta."
    )
    column_order: int = Field(
        ...,
        ge=0,
        description="Zero-based position in the target table schema."
    )


class TemplateColumnResponse(ORMBase):
    """Response shape for template columns."""

    id: UUID
    template_id: UUID
    column_name: str
    display_name: str | None
    data_type: str
    description: str | None
    is_included: bool
    is_pii: bool
    is_nullable: bool
    is_unique: bool
    column_order: int

File: frontend/src/types/template_column.ts
/**
 * TypeScript types for the TemplateColumn resource.
 *
 * Mirrors backend/app/schemas/template_column.py.
 *
 * A TemplateColumn defines a single column within a template:
 * its data type, whether it's PII, nullable, unique, etc.
 */

/**
 * Data types supported by the template column configuration.
 * Must match the SQL CHECK constraint in
 * sql/04_create_template_columns_table.sql.
 */
export type ColumnDataType =
  | 'STRING'
  | 'INTEGER'
  | 'BIGINT'
  | 'LONG'
  | 'DOUBLE'
  | 'DECIMAL'
  | 'BOOLEAN'
  | 'DATE'
  | 'TIMESTAMP'

/** Response shape returned by the backend. */
export interface TemplateColumn {
  id: string
  template_id: string

  column_name: string
  display_name: string | null
  data_type: ColumnDataType
  description: string | null

  is_included: boolean
  is_pii: boolean
  is_nullable: boolean
  is_unique: boolean
  column_order: number

  created_at: string
}

/**
 * Payload for creating a new template column.
 *
 * Optional fields marked with `?` may be omitted from the JSON
 * body. Backend applies defaults if missing.
 */
export interface TemplateColumnCreate {
  column_name: string
  display_name?: string | null
  data_type?: ColumnDataType
  description?: string | null
  is_included?: boolean
  is_pii?: boolean
  is_nullable?: boolean
  is_unique?: boolean
  column_order?: number
}